metamodel.zenta

```diff
@@ -2335,180 +2365,1111 @@ A role is assigned to human(s).
+              <bounds x="372" y="205" width="-1" height="-1"/>
+            </child>
+          </element>
+        </folder>
+        <folder name="3. Implementation pattern" id="262672f3-f9ce-4529-b897-182e7cbe849e">
+          <folder name="1. general rules" id="8b26c9c7-eb48-401e-a798-b6f12d219aac">
+            <element xsi:type="zenta:BasicObject" name="ancillary unit" id="75d5962e-721a-497b-932e-9a6dbe058d81" ancestor="behaviour">
+              <documentation>Ancillary units are put to the same package as the unit they support.
+
+* If the design or needs of testing mandates it, a DTO may have a Factory
+* an Entity must have a Repository
+* all DTO, Entity, Service and Controller units must have a Test Data
+* all Service and Controller units must have a Test Base
+* all Service and Controller units must have a Stub Provider
+* a legacy object may have a Factory and/or a Stub Provider and/or a Test Data. Those should be put in a package named 'legacy' under the main package of the project.
+
+</documentation>
+            </element>
+            <element xsi:type="zenta:BasicObject" name="code unit" id="478609a9-ef96-4d4b-b328-f982f1d0860d" ancestor="behaviour">
+              <documentation>code units do not contain data structures and constants
+those are in data structure and constant units</documentation>
+            </element>
+            <element xsi:type="zenta:BasicObject" name="complex constants and collections" id="8c8f5643-e1c9-4ca7-a18a-74b2ccf5cf99" ancestor="behaviour">
+              <documentation>* constants with complex structures (test Entity and DTO instances and mocks for legacy objects) and constant collections of constants are created by methods without parameters
+* these methods create a new instance of the constant every time they are called, using their constructor (not the factory)
+* the name of a method creating a complex constant should start with 'get'
+* the name of the method creating a collection of constants should start with the type of the collection, e.g. 'list'
+* the name of the method giving back a default instance ( a fully built instance meant to be used in the positive test case) is 'get()'
+* the name of the method giving back a collection containing only the default instance should be the name of the collection type, e.g. 'list()'
+* other methods must directly or indirectly use the get() method to deliver other instances by modifying the default instance.
+* the non-default methods must contain the name of the (most important) field modified right after the 'get' part. For example the method giving back an Entity with a zero id would be called getIdUninitialized()</documentation>
+            </element>
+            <element xsi:type="zenta:BasicObject" name="constant provider" id="861fcab2-587d-45f1-8587-1979180ee4f6" ancestor="behaviour">
+              <documentation>* constants can only be in constant provider units; move them there from other units
```

Comment:
As far as I remember we've been talking about constants used in a single place can be defined locally. Am I wrong?

Reply:
It have impacts on code checks. Let's discuss.